Sources/SwiftDocC/Model/Rendering/RenderSectionTranslator/DeclarationsSectionTranslator.swift

```diff
@@ -35,39 +37,284 @@ struct DeclarationsSectionTranslator: RenderSectionTranslator {
+            /// that is not shared. For example, a declaration such as `myFunc<T>(param: T)` will
+            /// have a text fragment `>(` before the parameter list, which is technically not
+            /// shared with a similar declaration `myFunc(param: Int)` which only has a `(`
+            /// fragment.
+            ///
+            /// Text should be broken up in three scenarios:
+            /// 1. Before a parenthesis or comma, as in the previous example,
+            /// 2. Before and after whitespace, to increase the chances of matching non-whitespace
+            ///    tokens with each other, and
+            /// 3. After a `?` character, to eagerly separate optional parameter types from their
+            ///    surrounding syntax.
+            ///
+            /// > Note: Any adjacent text fragments that are both shared or highlighted should
+            /// > be recombined after translation, to allow Swift-DocC-Render to correctly format
+            /// > Swift declarations into multiple lines. This is performed as part of
+            /// > `translateDeclaration(_:commonFragments:)` above.
+            func preProcessFragment(_ fragment: SymbolGraph.Symbol.DeclarationFragments.Fragment) -> [SymbolGraph.Symbol.DeclarationFragments.Fragment] {
+                guard fragment.kind == .text, !fragment.spelling.isEmpty else {
+                    return [fragment]
+                }
+                var textPartitions: [String] = []
+                var substringIndex = fragment.spelling.startIndex
+                var currentElement = fragment.spelling[substringIndex]
+
+                func areInSameChunk(_ currentElement: Character, _ nextElement: Character) -> Bool {
+                    if "(),".contains(nextElement) {
+                        // an open paren means we have a token like `>(` which should be split
+                        // a close paren means we have a token like ` = nil)` which should be split
+                        // a comma is similar to the close paren situation
+                        return false
+                    } else if currentElement.isWhitespace != nextElement.isWhitespace {
+                        // break whitespace into their own blocks
+                        return false
+                    } else if currentElement == "?" {
+                        // if we have a token like `?>` or similar we should break the fragment
+                        return false
+                    } else {
+                        return true
+                    }
+                }
+
+                // FIXME: replace this with `chunked(by:)` if we add swift-algorithms as a dependency
+                for (nextIndex, nextElement) in fragment.spelling.indexed().dropFirst() {
+                    if !areInSameChunk(currentElement, nextElement) {
+                        textPartitions.append(String(fragment.spelling[substringIndex..<nextIndex]))
```

Comment:
Is there any way to avoid allocating a new string here, for every chunk? Could `textPartitions` be an array of substrings or just string indices instead?

Reply:
The problem is that the `Fragment` type requires a fully allocated String instead of `any StringProtocol` or the like. If i kept `textPartitions` as an array of Substrings, i would just be allocating in the final `map` anyway.